package_python_function/reproducible_zip.py

```diff
@@ -0,0 +1,88 @@
+from __future__ import annotations
+
+import os
+import shutil
+import time
+from copy import copy
+from typing import TYPE_CHECKING
+from zipfile import ZipFile, ZipInfo
+
+if TYPE_CHECKING:
+    from typing import Tuple
+
+FILE_MODE = 0o644
+DIR_MODE = 0o755
+
+def date_time() -> Tuple[int, int, int, int, int, int]:
+    """Returns date_time value used to force overwrite on all ZipInfo objects. Defaults to
+    1980-01-01 00:00:00. You can set this with the environment variable SOURCE_DATE_EPOCH as an
+    integer value representing seconds since Epoch.
+    """
+    source_date_epoch = os.environ.get("SOURCE_DATE_EPOCH", None)
+    if source_date_epoch is not None:
+        return time.gmtime(int(source_date_epoch))[:6]
+    return (1980, 1, 1, 0, 0, 0)
+
+def clean_zip_info(zinfo: ZipInfo) -> ZipInfo:
+    """
+    Cleans the ZipInfo object, overwriting file-modified timestamps and file/directory permissions modes in write mode in order to create a reproducible ZIP archive.
+
+    Parameters:
+    zinfo (ZipInfo): A ZipInfo object from zipfile.ZipInfo.
+
+    Returns:
+    ZipInfo: The ZipInfo for the file, with the proper file permissions and date.
+    """
+    zinfo = copy(zinfo)
+    zinfo.date_time = date_time()
+    if zinfo.is_dir():
+        zinfo.external_attr = (0o40000 | DIR_MODE) << 16
+        zinfo.external_attr |= 0x10  # MS-DOS directory flag
+    else:
+        zinfo.external_attr = FILE_MODE << 16
+    return zinfo
+
+class ReproducibleZipFile(ZipFile):
```

Comment:
I think this can be done without hijacking ZipFile.  Can you try seeing if [ZipFile.open()](https://docs.python.org/3/library/zipfile.html#zipfile.ZipFile.open) will allow you to achieve this?  It should allow you to specify a ZipInfo directly, and then you get a file-like object you can write to using something like `shutil.copyfileobj`
Here is an example I found: https://stackoverflow.com/questions/8068180/adding-a-file-like-object-to-a-zip-file-in-python
Except you will pass your ZipInfo object instead of the filename string shown in that example

Reply:
amazing. Yea I'll totally give this a try!